src/wasm/wasm-binary.cpp

```diff
@@ -5645,6 +5688,33 @@ void WasmBinaryBuilder::visitCallRef(CallRef* curr) {
   curr->finalize(sig.results);
 }
 
+void WasmBinaryBuilder::visitLet(Block* curr) {
+  // A let is lowered into a block that contains the value, and we allocate
+  // locals as needed, which works as we remove non-nullability.
+
+  startControlFlow(curr);
+  // Get the output type.
+  curr->type = getType();
+  // Get the new local types.
+  auto varsBefore = currFunction->vars.size();
+  readVars();
+  auto numNewVars = currFunction->vars.size() - varsBefore;
+  // Assign the values into locals.
+  Builder builder(wasm);
+  std::vector<Index> absoluteIndexes;
+  for (Index i = 0; i < numNewVars; i++) {
+    auto* value = popNonVoidExpression();
+    auto absoluteIndex = varsBefore + i;
+    absoluteIndexes.push_back(absoluteIndex);
+    curr->list.push_back(builder.makeLocalSet(absoluteIndex, value));
+  }
+  // Read the body, with adjusted local indexes.
+  letStack.push_back(absoluteIndexes);
```

Comment:
Might as well move the vector into place rather than copying it.

```suggestion
  letStack.emplace_back(std::move(absoluteIndexes));
```